tests/crt/func_registry_test.cc

```diff
@@ -0,0 +1,223 @@
+            TVMFuncRegistry_GetByIndex(&registry, (tvm_function_index_t)0, NULL));
+}
+
+extern "C" {
+static int Foo(TVMValue* args, int* type_codes, int num_args, TVMValue* out_ret_value,
+               int* out_ret_tcode) {
+  return 0;
+}
+static int Bar(TVMValue* args, int* type_codes, int num_args, TVMValue* out_ret_value,
+               int* out_ret_tcode) {
+  return 0;
+}
+}
+
+// Matches the style of registry defined in generated C modules.
+const char* kBasicFuncNames = "\002Foo\0Bar\0";  // NOTE: final \0
+const TVMBackendPackedCFunc funcs[2] = {&Foo, &Bar};
+const TVMFuncRegistry kConstRegistry = {kBasicFuncNames, (const TVMBackendPackedCFunc*)funcs};
+
+TEST(FuncRegistry, ConstGlobalRegistry) {
+  tvm_function_index_t func_index = -1;
+  TVMBackendPackedCFunc func = nullptr;
+
+  // Foo
+  EXPECT_EQ(kBasicFuncNames[0], 2);
+  EXPECT_EQ(kBasicFuncNames[1], 'F');
+  EXPECT_EQ(kTvmErrorNoError, TVMFuncRegistry_Lookup(&kConstRegistry, "Foo", &func_index));
+  EXPECT_EQ(0, func_index);
+
+  EXPECT_EQ(kTvmErrorNoError, TVMFuncRegistry_GetByIndex(&kConstRegistry, func_index, &func));
+  EXPECT_EQ(func, &Foo);
+
+  // Bar
+  EXPECT_EQ(kTvmErrorNoError, TVMFuncRegistry_Lookup(&kConstRegistry, "Bar", &func_index));
+  EXPECT_EQ(1, func_index);
+
+  EXPECT_EQ(kTvmErrorNoError, TVMFuncRegistry_GetByIndex(&kConstRegistry, func_index, &func));
+  EXPECT_EQ(func, &Bar);
+
+  // Expected not found.
+  tvm_function_index_t prev_func_index = func_index;
+  EXPECT_EQ(kTvmErrorFunctionNameNotFound,
+            TVMFuncRegistry_Lookup(&kConstRegistry, "Baz", &func_index));
+  EXPECT_EQ(prev_func_index, func_index);
+
+  // Expected index out of range.
+  func = nullptr;
+  EXPECT_EQ(kTvmErrorFunctionIndexInvalid, TVMFuncRegistry_GetByIndex(&kConstRegistry, 2, &func));
+  EXPECT_EQ(func, nullptr);
+}
+
+/*! \brief Return a test function handle, with number repeating for all bytes in a void*. */
+static TVMBackendPackedCFunc TestFunctionHandle(uint8_t number) {
+  uintptr_t handle = 0;
+  for (size_t i = 0; i < sizeof(TVMBackendPackedCFunc); i++) {
+    handle |= ((uintptr_t)handle) << (8 * i);
+  }
+
+  return (TVMBackendPackedCFunc)handle;
+}
+
+TEST(MutableFuncRegistry, Create) {
+  char mem_buffer[kTvmAverageFuncEntrySizeBytes * 3];
+  // A substring used to create function names for testing.
+  const char* function_name_chars = "abcdefghijklmnopqrstuvwxyz";
+
+  // function_name_chars is used to produce 2 function names. The second one is expected to
+  // overfill `names`.
+  EXPECT_LT(kTvmAverageFuncEntrySizeBytes, strlen(function_name_chars) + 1);
+
+  for (unsigned int buf_size = 0; buf_size < 10 + 1 + sizeof(void*); buf_size++) {
+    EXPECT_EQ(kTvmErrorBufferTooSmall, TVMMutableFuncRegistry_Create(NULL, mem_buffer, buf_size));
+  }
+
+  for (unsigned int rem = 0; rem < kTvmAverageFuncEntrySizeBytes; rem++) {
+    // test_function name will be used to test overfilling.
+    char test_function_name[kTvmAverageFuncNameSizeBytes + 2 + rem];
```

Comment:
why `+ 2` here?

Reply:
its testing attempted over-filling, so + 1 for the overfill char and +1 for \0.